meson.build

```diff
@@ -0,0 +1,40 @@
+project('sage', 'c')
```

Comment:
I'd be more comfortable with this file being local to sage-conf_meson, so as not to create expectations that Sage-the-distribution will switch to meson

Reply:
The idea in the middle to long term is to support running `meson compile` from the root directory to compile sagelib with meson (using conda). For this the meson build file needs to be there (and its pretty much convention to have such a build file in the root). Sage-the-distro supporting meson would mean that there is a meson file in the build directory, or not?

Reply:
I'm not aware of such a plan, and I'd suggest that if you have a plan to use a new Issue in which you explain it.

From what you wrote above, it seems unlikely that it would be the same meson.build file that would be suitable for both purposes, so nothing is gained by putting it there.

Reply:
It's part of https://github.com/sagemath/sage/issues/34630. In the long-term you want to use of course meson-python (configured via pyproject.toml). But in the short to midterm its easier to just call meson directly. That's at least how scipy/numpy proceeded: https://github.com/scipy/scipy/pull/14847/files#diff-e42998d51257e6f84aa51ffa5f4ed1544cb12f97a94978c44f3bc281b5324d79R390-R500

Reply:
Well, #34630 has nothing to with stuff that belongs in SAGE_ROOT.

Reply:
Yes, devs usually start in the root.

Reply:
I have opened https://github.com/sagemath/sage/pull/36524. This should give you an idea on why its handy to have a top-level meson file. (It's very similar to having a top-level make file)

Reply:
Yes, looking at #36524 confirms: This is work on configuring the Sage library, not the Sage distribution.

Yes, obviously a top-level `meson.build` file is like a top-level `Makefile`. 

Our top-level Makefile is an element of the Sage distribution, not the Sage library.

So as I said, this new file does not belong in SAGE_ROOT.

Reply:
Sorry, but there are a lot of files in the root folder that are not related to the distribution. pyright config, tox ini, bootstrap conda, pipfile, contributing, gitpod, gitignore.

I think we also agree that sage's main purpose is sagelib, and not the distribution. So reserving the root folder for distribution seems to be conterproductive and giving unncessary prio to it.

It is natural to run `meson compile` from the root of the project. And since this workflow is easy to support, why shouldn't we?!

Reply:
> I think we also agree that sage's main purpose is sagelib, and not the distribution. So reserving the root folder for distribution seems to be counterproductive and giving unnecessary prio to it.

I don't disagree with this. And the work on `sage-conf_pypi` and `sage-conf_conda` certainly goes in the direction of "flipping the repository/distribution inside out", i.e., to make the configure script + the build system for SAGE_LOCAL just a detail of *one* implementation of sage-conf. See also #31662 and #30025.

But we cannot make such as structural change just casually on a ticket like this one:
- There's simply no technical need for the meson.build file of this PR here to be in the root.
- (Assuming that by "since this workflow is easy to support" you mean meson + conda for dependencies; correct me if you mean something else) There's no agreement in the community whatsoever that the project should do that.

Reply:
>But we cannot make such as structural change just casually on a ticket like this one:

There is no structural change in this PR. The meson build file provides an alternative to using autoconf, and is placed in the same folder as `configure.ac`. If you move the configure file to `pkgs/sage_conf` at some point, then the meson build file can be moved as well to `pkgs/sage_conf_meson`. Until then, for consistency, both (or none) should be in the root.

Reply:
> There is no structural change in this PR. The meson build file provides an alternative to using autoconf

No, it doesn't.

Reply:
I've now moved the config writing to a meson file in the pkgs folder. The meson file in the root is thus there only for convenience of developers (similar to how the tox ini in the root is directing a lot of stuff to the tox ini in src)

Reply:
How is it convenient and for what?

Reply:
To test the compilation/config gen. Meson supports eg multiple build directories, so you can play around with changing things while keeping the other generated version(s) around. Also calling meson directly is quicker than going through the pip install.

Reply:
So it's for compiling the Sage library, for the single use case of the all-conda install?

Reply:
Meson is somewhere between make and autoconf IMHO.

I don't see a point in insisting that its configs must be somewhere one has to find out from docs.

An alternative would be to have `meson` target in the top-level makefile.

Reply:
> Meson is somewhere between make and autoconf IMHO.

Yes, I think everyone here on this PR knows exactly what meson is.

Reply:
> So it's for compiling the Sage library, for the single use case of the all-conda install?

There is nothing conda-specific about the meson files, it's only that conda provides a nice fixed environment and thus is a good starting point. You can try to run `meson setup` outside of a conda environment and it should work (if you have all the necessary packages installed in the system).

Reply:
I'll explain and summarize once more: 
- What you just described ("it should work (if you have all the necessary packages installed in the system).") means that you are configuring / building the Sage library. 
- This is different from what our `configure.ac` and `Makefile` do. They configure / build the Sage distribution.
- There is no plan to switch the Sage distribution to using meson. 
- When people see a file `meson.build` next to `configure.ac` and `Makefile`, they are bound to think that using meson is an alternative for building the same thing. So this has the potential to mislead.

Reply:
> * When people see a file `meson.build` next to `configure.ac` and `Makefile`, they are bound to think that using meson is an alternative for building the same thing. So this has the potential to mislead.

So if you see a hammer and a saw next to each other you are bound to see them as alternatives for building the same thing?

Conceptually, `meson setup` does the same thing as `configure`: go through the declared dependencies and write out a config file. Similarly, `meson compile` will do the same thing as `make sage-all` (or whatever the correct target is for building sagelib). 

It's also not technically possible to move the meson file in the root to `src` since meson's `subdir` only walks down the tree.

Reply:
> Similarly, `meson compile` will do the same thing as `make sage-all` (or whatever the correct target is for building sagelib).

The target `make sage-all` does not exist. `make all`, the default target, builds the Sage distribution.

Reply:
> Sorry, but there are a lot of files in the root folder that are not related to the distribution. 
> [...] tox ini [...]

Actually `SAGE_ROOT/tox.ini` defines the portability tests of the Sage distribution.